pkg/controllers/terminating/v1alpha1/terminate.go

```diff
@@ -38,105 +38,62 @@ type Terminator struct {
+	// 1. Get pods on node
+	pods, err := t.getPods(ctx, node)
 	if err != nil {
-		return fmt.Errorf("listing draining nodes, %w", err)
+		return false, fmt.Errorf("listing pods for node %s, %w", node.Name, err)
 	}
-	// 2. Drain nodes
-	drained := []*v1.Node{}
-	for _, node := range draining {
-		// TODO: Check if Node should be drained
-		// - Disrupts PDB
-		// - Pods owned by controller object
-		// - Pod on Node can't be rescheduled elsewhere
-
-		// 2a. Get pods on node
-		pods, err := t.getPods(ctx, node)
-		if err != nil {
-			return fmt.Errorf("listing pods for node %s, %w", node.Name, err)
-		}
-		// 2b. Evict pods on node
-		empty := true
-		for _, p := range pods {
-			if !pod.IsOwnedByDaemonSet(p) {
-				empty = false
-				if err := t.coreV1Client.Pods(p.Namespace).Evict(ctx, &v1beta1.Eviction{
-					ObjectMeta: metav1.ObjectMeta{
-						Name: p.Name,
-					},
-				}); err != nil {
-					zap.S().Debugf("Continuing after failing to evict pods from node %s, %s", node.Name, err.Error())
-				}
+	// 2. Evict pods on node
+	empty := true
+	for _, p := range pods {
+		if !pod.IsOwnedByDaemonSet(p) {
+			empty = false
+			if err := t.coreV1Client.Pods(p.Namespace).Evict(ctx, &v1beta1.Eviction{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: p.Name,
+				},
+			}); err != nil {
+				zap.S().Debugf("Continuing after failing to evict pods from node %s, %s", node.Name, err.Error())
 			}
 		}
-		// 2c. If node is empty, add to list of nodes to delete
-		if empty {
-			drained = append(drained, node)
-		}
 	}
-	// 3. Delete empty nodes
-	if err := t.deleteNodes(ctx, drained); err != nil {
-		return fmt.Errorf("deleting %d nodes, %w", len(drained), err)
-	}
-	return nil
+	return empty, nil
 }
 
-// deleteNode uses a cloudprovider-specific delete to delete a set of nodes
-func (t *Terminator) deleteNodes(ctx context.Context, nodes []*v1.Node) error {
-	// 1. Delete node in cloudprovider's instanceprovider
-	if err := t.cloudProvider.Terminate(ctx, nodes); err != nil {
+// terminateNode terminates the node then removes the finalizer to delete the node
+func (t *Terminator) terminateNode(ctx context.Context, n *v1.Node) error {
+	if err := t.cloudProvider.Terminate(ctx, n); err != nil {
 		return fmt.Errorf("terminating cloudprovider instance, %w", err)
 	}
-	// 2. Delete node in APIServer
-	for _, node := range nodes {
-		if err := t.kubeClient.Delete(ctx, node); err != nil {
-			zap.S().Debugf("Continuing after failing to delete node %s, %s", node.Name, err.Error())
+	zap.S().Infof("Terminated instance %s", n.Name)
+	// 2. Remove finalizer from node in APIServer
+	for i, finalizer := range n.Finalizers {
+		if finalizer == provisioning.KarpenterFinalizer {
+			persisted := n.DeepCopy()
+			node.RemoveFinalizer(n, i)
+			if err := t.kubeClient.Patch(ctx, n, client.MergeFrom(persisted)); err != nil {
+				return fmt.Errorf("removing finalizer from node %s, %w", n.Name, err)
```

Comment:
What actually happens in this case (later on)? Does it all get re-driven somehow? (Especially since the cloud instance probably is gone now, so (possibly?) some of the earlier calls will start failing (although, see other comments).

Reply:
This happens right after the instance is terminated. If this patch fails, it will have ~1 hour to succeed to remove the finalizer, since that's when EC2 will then return a failure here. Does this answer your question?